File: 백준/Gold/10986. 나머지 합/나머지 합.java
import java.util.*;
import java.io.BufferedReader;
import java.io.InputStreamReader;

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());
        
        long[] parSum = new long[N];
        long[] res = new long[M];

        long ans = 0;
        st = new StringTokenizer(br.readLine());
        parSum[0] = Integer.parseInt(st.nextToken());
        for(int i = 1; i < N; ++i){
            parSum[i] = parSum[i - 1] + Integer.parseInt(st.nextToken());
        }

        for(int i = 0; i < N; ++i){
            if((int)(parSum[i] % M) == 0)
                ans++;

            res[(int)(parSum[i] % M)]++;
            
        }

        for(int i = 0; i < M; ++i){
            if(res[i] > 1)
            ans += (res[i] * (res[i] - 1)) / 2;
        }
        System.out.println(ans);
    }
}

File: 백준/Gold/10986. 나머지 합/나머지 합.py
import sys
input = sys.stdin.readline

N, M = map(int, input().split())
A = list(map(int, input().split()))
parSum = [0 for _ in range(N)]
parSum[0] = A[0]

for i in range(1, N):
    parSum[i] = parSum[i - 1] + A[i]

res = [0 for _ in range(M)]
answer = 0
for i in range(N):
    if parSum[i] % M == 0:
        answer += 1
    res[parSum[i] % M] += 1

for i in range(M):
    if res[i] > 0:
        answer += (res[i] * (res[i] - 1)) // 2
print(answer)